File: blobs/howto/JavaScript/NodeJS-v12/dev-guide/copy-blob.js
// Azure Storage dependency
const {
  BlobServiceClient,
  BlobLeaseClient,
  BlockBlobClient,
  BlobSASPermissions,
  generateBlobSASQueryParameters,
  SASProtocol
} = require('@azure/storage-blob');

// Azure authentication for credential dependency
const { DefaultAzureCredential } = require('@azure/identity');

// For development environment - include environment variables from .env
require("dotenv").config();

// TODO: Replace with your actual storage account name
const accountNameSrc = "<storage-account-name-src>";
const accountNameDest = "<storage-account-name-dest>";

async function main() {
  // create service client from DefaultAzureCredential
  const blobServiceClient = new BlobServiceClient(
    `https://${accountNameSrc}.blob.core.windows.net`,
    new DefaultAzureCredential()
  );
  const blobServiceClientDest = new BlobServiceClient(
    `https://${accountNameDest}.blob.core.windows.net`,
    new DefaultAzureCredential()
  );

  const sourceBlob = blobServiceClient
    .getContainerClient("source-container")
    .getBlockBlobClient("sample-blob.txt");

  const destinationBlob = blobServiceClientDest
    .getContainerClient("destination-container")
    .getBlockBlobClient("sample-blob-copy.txt");

  copyAcrossStorageAccountsAsync(sourceBlob, destinationBlob, blobServiceClient);

  const sourceURL = "<source-url>";
  const destBlob = blobServiceClient
    .getContainerClient("sample-container")
    .getBlockBlobClient("sample-blob-copy.txt");

  copyFromExternalSource(sourceURL, destBlob);
}

// <Snippet_copy_from_azure_async>
async function copyAcrossStorageAccountsAsync(sourceBlob, destinationBlob, blobServiceClient) {
  // Lease the source blob to prevent changes during the copy operation
  const sourceBlobLease = new BlobLeaseClient(sourceBlob);

  // Create a SAS token that's valid for 1 hour
  const sasToken = await generateUserDelegationSAS(sourceBlob, blobServiceClient);
  const sourceBlobSASURL = sourceBlob.url + "?" + sasToken;

  try {
    await sourceBlobLease.acquireLease(-1);

    // Start the copy operation and wait for it to complete
    const copyPoller = await destinationBlob.beginCopyFromURL(sourceBlobSASURL);
    await copyPoller.pollUntilDone();
  } catch (error) {
    // Handle the exception
  } finally {
    // Release the lease once the copy operation completes
    await sourceBlobLease.releaseLease();
  }
}

async function generateUserDelegationSAS(sourceBlob, blobServiceClient) {
  // Get a user delegation key for the Blob service that's valid for 1 hour, as an example
  const delegationKeyStart = new Date();
  const delegationKeyExpiry = new Date(Date.now() + 3600000);
  const userDelegationKey = await blobServiceClient.getUserDelegationKey(
    delegationKeyStart,
    delegationKeyExpiry
  );

  // Create a SAS token that's valid for 1 hour, as an example
  const sasTokenStart = new Date();
  const sasTokenExpiry = new Date(Date.now() + 3600000);
  const blobName = sourceBlob.name;
  const containerName = sourceBlob.containerName;
  const sasOptions = {
    blobName,
    containerName,
    permissions: BlobSASPermissions.parse("r"),
    startsOn: sasTokenStart,
    expiresOn: sasTokenExpiry,
    protocol: SASProtocol.HttpsAndHttp
  };

  const sasToken = generateBlobSASQueryParameters(
    sasOptions,
    userDelegationKey,
    blobServiceClient.accountName
  ).toString();

  return sasToken.toString();
}
// </Snippet_copy_from_azure_async>

// <Snippet_copy_blob_external_source_async>
async function copyFromExternalSource(sourceURL, destinationBlob) {
  const copyPoller = await destinationBlob.beginCopyFromURL(sourceURL);
  await copyPoller.pollUntilDone();
}
// </Snippet_copy_blob_external_source_async>

// <Snippet_check_copy_status_async>
async function checkCopyStatus(destinationBlob) {
  const properties = await destinationBlob.getProperties();
  console.log(properties.copyStatus);
}
// </Snippet_check_copy_status_async>

// <Snippet_abort_copy_async>
async function abortCopy(destinationBlob) {
  const properties = await destinationBlob.getProperties();

  // Check the copy status and abort if pending
  if (properties.copyStatus === "pending") {
    await destinationBlob.abortCopyFromURL(properties.copyId);
  }
}
// </Snippet_abort_copy_async>

main()
    .then(() => console.log('done'))
    .catch((ex) => console.log(ex.message));

File: blobs/howto/dotnet/BlobDevGuideBlobs/CopyBlob.cs
using Azure;
using Azure.Storage.Blobs;
using Azure.Storage.Blobs.Models;
using Azure.Storage.Blobs.Specialized;
using Azure.Storage.Sas;

namespace BlobDevGuide
{
    class CopyBlob
    {
        //-------------------------------------------------
        // Check blob copy status
        //-------------------------------------------------
        // <Snippet_CheckStatusCopyBlob>
        public static async Task CheckCopyStatusAsync(CopyFromUriOperation copyOperation)
        {
            // Check for the latest status of the copy operation
            Response response = await copyOperation.UpdateStatusAsync();

            // Parse the response to find x-ms-copy-status header
            if (response.Headers.TryGetValue("x-ms-copy-status", out string value))
                Console.WriteLine($"Copy status: {value}");
        }
        // </Snippet_CheckStatusCopyBlob>

        // <Snippet_CopyWithinAccount_CopyBlob>
        //-------------------------------------------------
        // Copy a blob from the same storage account
        //-------------------------------------------------
        public static async Task CopyWithinStorageAccountAsync(
            BlobClient sourceBlob,
            BlockBlobClient destinationBlob)
        {
            // Lease the source blob to prevent changes during the copy operation
            BlobLeaseClient sourceBlobLease = new(sourceBlob);

            try
            {
                await sourceBlobLease.AcquireAsync(BlobLeaseClient.InfiniteLeaseDuration);

                // Start the copy operation and wait for it to complete
                CopyFromUriOperation copyOperation = 
                    await destinationBlob.StartCopyFromUriAsync(sourceBlob.Uri);
                await copyOperation.WaitForCompletionAsync();
            }
            catch (RequestFailedException ex)
            {
                // Handle the exception
            }
            finally
            {
                // Release the lease once the copy operation completes
                await sourceBlobLease.ReleaseAsync();
            }
        }
        // </Snippet_CopyWithinAccount_CopyBlob>

        // <Snippet_CopyAcrossAccounts_CopyBlob>
        //-------------------------------------------------
        // Copy a blob from a different storage account
        //-------------------------------------------------
        public static async Task CopyAcrossStorageAccountsAsync(
            BlobClient sourceBlob,
            BlockBlobClient destinationBlob)
        {
            // Lease the source blob to prevent changes during the copy operation
            BlobLeaseClient sourceBlobLease = new(sourceBlob);

            // Create a Uri object with a SAS token appended - specify Read (r) permissions
            Uri sourceBlobSASURI = await GenerateUserDelegationSAS(sourceBlob);

            try
            {
                await sourceBlobLease.AcquireAsync(BlobLeaseClient.InfiniteLeaseDuration);

                // Start the copy operation and wait for it to complete
                CopyFromUriOperation copyOperation = await destinationBlob.StartCopyFromUriAsync(sourceBlobSASURI);
                await copyOperation.WaitForCompletionAsync();
            }
            catch (RequestFailedException ex)
            {
                // Handle the exception
            }
            finally
            {
                // Release the lease once the copy operation completes
                await sourceBlobLease.ReleaseAsync();
            }
        }

        async static Task<Uri> GenerateUserDelegationSAS(BlobClient sourceBlob)
        {
            BlobServiceClient blobServiceClient =
                sourceBlob.GetParentBlobContainerClient().GetParentBlobServiceClient();

            // Get a user delegation key for the Blob service that's valid for 1 day
            UserDelegationKey userDelegationKey =
                await blobServiceClient.GetUserDelegationKeyAsync(DateTimeOffset.UtcNow,
                                                                  DateTimeOffset.UtcNow.AddDays(1));

            // Create a SAS token that's also valid for 1 day
            BlobSasBuilder sasBuilder = new BlobSasBuilder()
            {
                BlobContainerName = sourceBlob.BlobContainerName,
                BlobName = sourceBlob.Name,
                Resource = "b",
                StartsOn = DateTimeOffset.UtcNow,
                ExpiresOn = DateTimeOffset.UtcNow.AddDays(1)
            };

            // Specify read permissions for the SAS
            sasBuilder.SetPermissions(BlobSasPermissions.Read);

            // Add the SAS token to the blob URI
            BlobUriBuilder blobUriBuilder = new BlobUriBuilder(sourceBlob.Uri)
            {
                // Specify the user delegation key
                Sas = sasBuilder.ToSasQueryParameters(userDelegationKey,
                                                      blobServiceClient.AccountName)
            };

            return blobUriBuilder.ToUri();
        }
        // </Snippet_CopyAcrossAccounts_CopyBlob>

        // <Snippet_CopyFromExternalSource_CopyBlob>
        //-------------------------------------------------
        // Copy a blob from an external source
        //-------------------------------------------------
        public static async Task CopyFromExternalSourceAsync(
            string sourceLocation,
            BlockBlobClient destinationBlob)
        {
            Uri sourceUri = new(sourceLocation);

            // Start the copy operation and wait for it to complete
            CopyFromUriOperation copyOperation = await destinationBlob.StartCopyFromUriAsync(sourceUri);
            await copyOperation.WaitForCompletionAsync();
        }
        // </Snippet_CopyFromExternalSource_CopyBlob>

        //-------------------------------------------------
        // Abort a blob copy operation
        //-------------------------------------------------
        // <Snippet_AbortBlobCopy>
        public static async Task AbortBlobCopyAsync(
            CopyFromUriOperation copyOperation,
            BlobClient destinationBlob)
        {
            // Check for the latest status of the copy operation
            Response response = await copyOperation.UpdateStatusAsync();

            // Parse the response to find x-ms-copy-status header
            if (response.Headers.TryGetValue("x-ms-copy-status", out string value))
            {
                if (value == "pending")
                {
                    await destinationBlob.AbortCopyFromUriAsync(copyOperation.Id);
                    Console.WriteLine($"Copy operation {copyOperation.Id} aborted");
                }
            }
        }
        // </Snippet_AbortBlobCopy>        
    }
}
